contracts/src/registry/PermissionedRegistry.sol

```diff
@@ -2,55 +2,264 @@
+    uint96 public constant FLAGS_MASK = 0xffffffff; // 32 bits
+
+    constructor(IRegistryDatastore _datastore, RegistryMetadata _metadata) BaseRegistry(_datastore) MetadataMixin(_metadata) {
+        _grantRoles(ROOT_RESOURCE, ALL_ROLES, _msgSender());
+
+        if (address(_metadata) == address(0)) {
+            _updateMetadataProvider(new SimpleRegistryMetadata());
+        }
+    }
 
-    constructor(IRegistryDatastore _datastore) BaseRegistry(_datastore) {
+    function uri(uint256 tokenId ) public view override returns (string memory) {
+        return tokenURI(tokenId);
     }
 
-    function _setFlags(uint256 tokenId, uint96 _flags)
-        internal
-        withSubregistryFlags(tokenId, FLAG_FLAGS_LOCKED, 0)
-        returns(uint96 newFlags)
+    function ownerOf(uint256 tokenId)
+        public
+        view
+        virtual
+        override(ERC1155Singleton, IERC1155Singleton)
+        returns (address)
     {
-        (address subregistry, uint96 oldFlags) = datastore.getSubregistry(tokenId);
-        newFlags = oldFlags | (_flags & FLAGS_MASK);
-        if (newFlags != oldFlags) {
-            datastore.setSubregistry(tokenId, subregistry, newFlags);
+        (, uint96 oldFlags) = datastore.getSubregistry(tokenId);
+        uint64 expires = _extractExpiry(oldFlags);
+        if (expires < block.timestamp) {
+            return address(0);
         }
+        return super.ownerOf(tokenId);
     }
 
-    function setSubregistry(uint256 tokenId, IRegistry registry)
+    function register(string calldata label, address owner, IRegistry registry, address resolver, uint96 flags, uint256 roleBitmap, uint64 expires, string memory _uri)
+        public
+        onlyRootRoles(ROLE_REGISTRAR)
+        returns (uint256 tokenId)
+    {
+        tokenId = _computeTokenId(label, flags);
+        flags = _computeFlags(flags, expires);
+
+        (, uint96 oldFlags) = datastore.getSubregistry(tokenId);
+        uint64 oldExpiry = _extractExpiry(oldFlags);
+        if (oldExpiry >= block.timestamp) {
+            revert NameAlreadyRegistered(label);
+        }
+
+        if (expires < block.timestamp) {
+            revert CannotSetPastExpiration(expires);
+        }
+
+        // if there is a previous owner, burn the token
+        address previousOwner = super.ownerOf(tokenId);
+        if (previousOwner != address(0)) {
+            _burn(previousOwner, tokenId, 1);
+        }
+
+        _mint(owner, tokenId, 1, "");
+        _grantRoles(tokenIdResource(tokenId), roleBitmap, owner);
+
+        datastore.setSubregistry(tokenId, address(registry), flags);
+        datastore.setResolver(tokenId, resolver, flags);
+
+        if (bytes(_uri).length > 0) {
+            metadataProvider.setTokenUri(tokenId, _uri);
+            emit URI(_uri, tokenId);
+        }
+        
+        emit NewSubname(tokenId, label);
+
+        return tokenId;
+    }
+
+    function setTokenObserver(uint256 tokenId, address _observer) external onlyTokenOwner(tokenId) {
```

Comment:
Should there be a permission to set a token observer?

Reply:
Well, only token owner can set it. That should do, right?

Reply:
Right, I'm suggesting it would be more flexible if that's a right that can be delegated like others here.